### DISCUSSION TITLE
Compiling into a binary

### DISCUSSION THREAD

**Discussion by adaliszk** [zavoloklom/docker-compose-linter#37] [Ideas]
To minimize the image size and enable tooling patterns where you maintain images for each workflow stage, it would be very handy to have a single binary that works in most Linux environments. While Node has a few options here, in my experience it has been quite the pain to actually combine the node binary with a slimmed down distribution and get it to be signed correctly so it can be tricky to introduce it, but the earlier its done the easier to keep it maintained.

What do you think?

**Comment by zavoloklom** [comment]
Hi @adaliszk!
Personally I thought that [docker image](https://hub.docker.com/r/zavoloklom/dclint) would solve it.

Could you please add a bit more information about your suggestion, maybe you could provide some examples?

**Reply by adaliszk** [reply]
The docker image is OK to distribute, but with the current setup where the node executes a js file, it is not portable to other images. It inherently requires any other image to install a compatible node version and copy over all the dependencies alongside the actual script.

In my case, we collect tools into specialized images that we build specifically with collapsed layers so that our CI/CD runners do not need to store many layers and can easily cache the intended tools with the least amount of image size.

For example, an excerpt for our containerization tools image looks like this:
```Dockerfile
FROM ghcr.io/hadolint/hadolint:v2.12.0 as hadolint
FROM bitnami/minideb:bookworm as runtime
RUN install_packages jq git git-lfs

COPY --from=hadolint /bin/hadolint /bin/hadolint

# Allow GitLab CI to decide on the entrypoint via asking the shell environment for an executable
ENTRYPOINT [ "" ]
```

So far, every tool we collected has been compiled into single binaries, which allows us to copy them into the image.

Now, with `dclint`, we would need to not only copy the script and the node_modules but also install an entire node runner, effectively increasing our image size from ~100MB to about ~200MB, which is a really large difference for the benefit of checking a YAML file.

While for me this tool helps a lot, I cannot convince others that something that they never checked beyond syntax (with the built-in linter) is wroth the effort when it doubles the respective docker image size.

**Reply by adaliszk** [reply]
There are some options for making the tool a binary, but I don't know which one you would like.

The most ideal would be to compile node and drop all modules not needed for this project to minimize its environment size and embed the script and its tree-shake dependencies into one binary. However, in my experience, this isn't easy to do, mainly because of complexities that are not explained well on the documentation page.

Alternative options include Deno or Bun, which both have a much simpler way to compile an entire project into a binary and embed their runtime. Additionally, there is always the option to 🦀😄

**Reply by zavoloklom** [reply]
I'll search for solution for Nodejs, don't want to switch language right now. 
Also if you know something about it or even can provide some examples of this kind of node libs - fell free to mention it here.

**Reply by zavoloklom** [reply]
Hi @adaliszk,  
I managed to build the binary by rewriting the build to rollup and using Single Executable Applications ([https://nodejs.org/docs/latest-v20.x/api/single-executable-applications.html](https://nodejs.org/docs/latest-v20.x/api/single-executable-applications.html)). The result is around 130MB, but it still depends on a couple of C++ libraries. To run it on a scratch image, I need to set it up like this:

```dockerfile
# some build stages

FROM scratch

ENV NODE_NO_WARNINGS=1

COPY --from=builder /bin/dclint /bin/dclint

COPY --from=builder "/usr/lib/libstdc++.so.6" "/usr/lib/libstdc++.so.6"
COPY --from=builder /usr/lib/libgcc_s.so.1 /usr/lib/libgcc_s.so.1
COPY --from=builder /lib/ld-musl-aarch64.so.1 /lib/ld-musl-aarch64.so.1
COPY --from=builder /lib/libc.musl-aarch64.so.1 /lib/libc.musl-aarch64.so.1

WORKDIR /app

ENTRYPOINT ["/bin/dclint"]
```

To run on alpine:

```Dockerfile
# some build stages

FROM alpine:3.19

ENV NODE_NO_WARNINGS=1

RUN apk add --no-cache libstdc++

COPY --from=builder /bin/dclint /bin/dclint

WORKDIR /app

ENTRYPOINT ["/bin/dclint"]
```

What do you think? If I upload it somewhere, would you be able to test it? What do you prefer - binary or docker image or both?

**Comment by zavoloklom** [comment]
@adaliszk Hi! I add this feature to a new version - https://github.com/zavoloklom/docker-compose-linter/releases/tag/v2.0.0
You can see binary files in this release. Let me know if something wrong with a new issue.
